src/parallel.cpp

```diff
@@ -17,38 +17,49 @@
 
 #include <boost/thread/thread.hpp>
 
-#define MAX_SCRIPT_CHECK_QUEUES 4
-
-static CCheckQueue<CScriptCheck> scriptcheckqueue1(128);
-static CCheckQueue<CScriptCheck> scriptcheckqueue2(128);
-static CCheckQueue<CScriptCheck> scriptcheckqueue3(128);
-static CCheckQueue<CScriptCheck> scriptcheckqueue4(128);
-
 using namespace std;
 
+static const unsigned int nScriptCheckQueues = 4;
+std::unique_ptr<CParallelValidation> PV;
 
-void AddScriptCheckThreads(int i, CCheckQueue<CScriptCheck> *pqueue)
+static void HandleBlockMessageThread(CNode *pfrom, const std::string strCommand, const CBlock block, const CInv inv);
+
+static void AddScriptCheckThreads(int i, CCheckQueue<CScriptCheck> *pqueue)
 {
     ostringstream tName;
     tName << "bitcoin-scriptchk" << i;
     RenameThread(tName.str().c_str());
     pqueue->Thread();
 }
 
-void AddAllScriptCheckQueuesAndThreads(int nScriptCheckThreads, boost::thread_group *threadGroup)
+CParallelValidation::CParallelValidation(int threadCount, boost::thread_group *threadGroup)
+    : semaphore(nScriptCheckQueues)
 {
-    CCheckQueue<CScriptCheck> *pqueue[MAX_SCRIPT_CHECK_QUEUES] = {
-        &scriptcheckqueue1, &scriptcheckqueue2, &scriptcheckqueue3, &scriptcheckqueue4};
+    if (threadCount <= 1)
+        threadCount = 0;
+    else if (threadCount > MAX_SCRIPTCHECK_THREADS)
+        threadCount = MAX_SCRIPTCHECK_THREADS;
+    nThreads = threadCount;
+
+    LogPrintf("Using %d threads for script verification\n", threadCount);
 
-    for (int i = 0; i < MAX_SCRIPT_CHECK_QUEUES; i++)
+    while (QueueCount() < nScriptCheckQueues)
     {
-        allScriptCheckQueues.Add(pqueue[i]);
-        for (int j = 0; j < nScriptCheckThreads; j++)
-            threadGroup->create_thread(boost::bind(&AddScriptCheckThreads, i + 1, pqueue[i]));
+        std::unique_ptr<CCheckQueue<CScriptCheck> > queue(new CCheckQueue<CScriptCheck>(128));
+
+        for (unsigned int i = 0; i < nThreads; i++)
+            threadGroup->create_thread(boost::bind(&AddScriptCheckThreads, i + 1, queue.get()));
+
+        vQueues.push_back(std::move(queue));
```

Comment:
It would be easy to make nScriptCheckQueues an argument to the constructor, but previously it was a constant (albeit as a macro) and I just preserved that.  I can make it an argument in a follow-up if you think that's an improvement.  There is no command line argument to specify it at present though so we'd just be passing 4 from init.cpp.